### ISSUE TITLE
Frames with balanced cutouts / flow around frames

### ISSUE THREAD

**Issue by alerque** [sile-typesetter/sile#1572]
We finally have decent paragraph shaping to do dropcaps, but currently our method is tightly linked to *paragraphs*. Sometimes the shaping needed in more closely related to frames than the content in them.

Here is an example from the [NET Full Notes Edition](https://netbible.com/buy-now/net-bible-full-notes-edition/):

![image](https://user-images.githubusercontent.com/173595/189478809-7e1a612f-0f58-4a04-be25-d6ed6311d5e3.png)

**Comment by Omikhleia** [comment]
Oh yes! I wondered about it too (not necessarily for text "insets" as here, but also for images, etc.). Such layouts are also pretty complex with TeX/LaTeX... and fragile (notwithstanding the question of balanced frames...). And while looking challenging, the example you gave also has many flaws, on a purely typographic angle (my eyes bleed at several details, e.g. the top and bottom line alignments are.. awful).

**Comment by alerque** [comment]
Yes I looked through some of the other pages visible online and there are quite a few other problems. Looking at the result I think I can guess about a lot of the tooling problems they faced making the project happen at all. Given what they crammed in I'd say they did a pretty good job, but it also leaves room for improvement.

----

I'm guessing the way for us to implement this (and many similar things like images as you note) is probably to use the existing paragraph shaping stuff you wrong with a custom callback function that knows it's position relative to the frame and goes out and checks other frames/objects/shapes and returns appropriate setbacks for that line. With this sort of scheme one could even wrap around arbitrary shapes like SVG outlines by having a frame that was the bounding box of an SVG shape, but the inside the frame use a callback that did a path intersection to find the relevant bit of the shape to align to.

The biggest issue I see right now is that this would not play nice with our vertical spacing mechanism. It might work with fixed or grid based alignments, but anything flexible where different lines might have different leading would be a nightmare. Recalculating the leading could change the target line length and hence potentially trigger a reflow that requires new leading calcs.
